numpy/array_api/_array_object.py

```diff
@@ -292,100 +290,122 @@ def _validate_index(key, shape):
+                else:
+                    f_range = f"between -{side} and {side - 1} (or None)"
+                if i.start is not None:
+                    try:
+                        start = operator.index(i.start)
+                    except TypeError:
+                        pass  # handled by ndarray
+                    else:
+                        if not (-side <= start <= side):
+                            raise IndexError(
+                                f"Slice {i} contains {start=}, but should be "
+                                f"{f_range} for an axis of size {side} "
+                                "(out-of-bounds starts are not specified in "
+                                "the Array API)"
+                            )
+                if i.stop is not None:
+                    try:
+                        stop = operator.index(i.stop)
+                    except TypeError:
+                        pass  # handled by ndarray
+                    else:
+                        if not (-side <= stop <= side):
+                            raise IndexError(
+                                f"Slice {i} contains {stop=}, but should be "
+                                f"{f_range} for an axis of size {side} "
+                                "(out-of-bounds stops are not specified in "
+                                "the Array API)"
+                            )
+            elif isinstance(i, Array):
+                if i.dtype in _boolean_dtypes and len(_key) != 1:
+                    assert isinstance(key, tuple)  # sanity check
                     raise IndexError(
-                        "Non-zero dimensional integer array indices are not allowed in the array API namespace"
+                        f"Single-axes index {i} is a boolean array and "
+                        f"{len(key)=}, but masking is only specified in the "
+                        "Array API when the array is the sole index."
                     )
-            return key._array
-        elif key is Ellipsis:
-            return key
-        elif key is None:
-            raise IndexError(
-                "newaxis indices are not allowed in the array API namespace"
-            )
-        try:
-            key = operator.index(key)
-            if shape is not None and len(shape) > 1:
+                elif i.dtype in _integer_dtypes and i.ndim != 0:
+                    raise IndexError(
+                        f"Single-axes index {i} is a non-zero-dimensional "
+                        "integer array, but advance integer indexing is not "
+                        "specified in the Array API."
+                    )
+            elif isinstance(i, tuple):
                 raise IndexError(
-                    "Multidimensional arrays must include an index for every axis or use an ellipsis"
+                    f"Single-axes index {i} is a tuple, but nested tuple "
+                    "indices are not specified in the Array API."
```

Comment:
This is maybe a little misleading, since this doesn't make sense in NumPy either.

Reply:
I guess NumPy does sometimes allow a tuple to represent an integer array. But it's deprecated.

Reply:
> I guess NumPy does sometimes allow a tuple to represent an integer array. But it's deprecated.

Ah I didn't realise this was deprecated. I think it's better to be safe here, especially considering no warning is given when you use nested tuples for integer indexing e.g. `np.arange(5)[((0, 1),)]`.

Reply:
That's actually the reason why we handle it here at all. If NumPy errored, we could just pass it through to ndarray.

Reply:
The deprecation period is over in NumPy.  NumPy parses any tuple as a multidimensional index (or tries to), and _within_ a multidimensional index will convert any sequence to an array/advanced index.

Reply:
Maybe I'm misremembering what is deprecated. I would have assumed that the ideal behavior would be for a tuple inside of a tuple to be an error, and only treat a list as an array.

Reply:
tuples are special.  Once inside a tuple (n-dimensional index), NumPy allows any "array-like", and tuples are considered array-like as well after all.
That is legitimate to not allow all NumPy array-likes of course, but I think it makes sense for NumPy, even if it may be a bit surprising if you write `arr[(1, 2),]`.

Reply:
I half suspsect this whole branch is unreachable anyway with the "check tuple first" reorganization.

In any case, whether or not this could be refactored more, I think we should put this in once @asmeurer did a final review.